DOCS SilverStripe RFC process, template and links to past RFCs

camfindlay opened the pull request:
Pulls together the process and ideas for having a RFC process outlined in https://groups.google.com/forum/#!topic/silverstripe-dev/5IEGRHNgToU and draws from the lessons learned in the pilot RFC at #3792 

Review and improvements welcome (please PR direct to my branch).

@mateusz you might want to comment/improve based on your learnings.

chillu commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+This is part of the SilverStripe core decision-making process and addresses the gap between the idea phase and the pull request submission (and merge).
+
+The rationale behind this process is to:
+ * Improve visibility on how the decisions made, producing a more open community mood
```
I'd replace with "Encourage visibility on decision making". "community mood" sounds weird.

chillu commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+ * Improve visibility on how the decisions made, producing a more open community mood
+ * Clarity on what the proposal is, its rationale and impact
+ * Reduce unnecessary work when coding large features is done without community and core-committers buy-in
+ * Improved likelihood of an optimal solution being merged - the pressure is much higher to approve the change if the coding has already been done (it’s much easier to make adjustments in the design phase)
```
Its unclear if "pressure much higher" is a positive or negative sentiment here. I'd just keep the first sentence, the second one is already contained in the previous bullet point

chillu commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+## What next?
+The RFC will be raised and discussed by the core committers in the monthly Google Hangout sessions, a vote for accepting the RFC will be taken requiring a majority vote (with at least a quorum of more than half of the core committers present).
+
+Once approved this means that if a pull request meeting the idea set out in the RFC was raised that it would be merged (pending the usual code peer review).
```
Mention when the Github issue will be closed - I assume when the code is merged (not waiting until a release)

chillu commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+## RFC Archive
+ RFC # | Proposal | Status
+------ |----------|-------
+RFC-1  | [Asset abstraction](https://github.com/silverstripe/silverstripe-framework/issues/3792) | Approved
```
That list doesn't really scale - it'll go stale very quickly. I'd suggest we simply have a search for the term "RFC". Or a new Github issue tag, but we kinda already have too many of these. Don't think there's a need to formalise the "approved" state beyond the metadata contained in Markdown already.

camfindlay replied:
Search brings up a lot of irrelevant things https://github.com/silverstripe/silverstripe-framework/search?q=RFC-&state=open&type=Issues&utf8=%E2%9C%93 - I can drop the status and think perhaps it couldn't hurt to have a master list of the proposed RFC's that follow the set process. If we are going to have ref numbers on them people will need to find out what the next relevant number is right? Could be stored in the current stable version documentation as I'm doing here?

mateusz commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+ * **Impact** - How will this change potentially impact on SilverStripe core? The good and the bad.
+
+###Submitting
+Once complete submit the RFC as a GitHub issue, in markdown format and include the ID number of the RFC based on the current list of existing proposals.
```
There are two problems I've encountered with github issues:
* inability to do TOCs - the links to chapters are not working so it's quite hard to navigate if the document is much more substantial
* inability to store images (I've pushed them to i.imgur.com, but who knows when they will remove it - in a month? in a year?).

camfindlay replied:
Have you got a recommended place we should put RFCs instead?

mateusz commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+### Template
+The following heading can act as a template to starting your RFC.
+ * **Introduction** - include a reference #, title, author
+ * **Motivation** - why this is a good idea
```
I'd say the first two items should be standardised metadata, and the second item before the motivation should be a "Purpose and outcome". Otherwise the document lacks the overview statement. I've added such a section to [RFC-1](https://github.com/silverstripe/silverstripe-framework/issues/3792) now.

Maybe something like this:

* Metadata - standardised header containing at least the Author(s), Status and Version fields.
* Purpose and outcome - the purpose of this document, and the expected outcome.
* ...and then continued starting from Motivation

camfindlay replied:
Fancy adding your changes via pull request to https://github.com/camfindlay/silverstripe-framework/tree/rfc ? ;)

mateusz commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+ * Reduce unnecessary work when coding large features is done without community and core-committers buy-in
+ * Improved likelihood of an optimal solution being merged - the pressure is much higher to approve the change if the coding has already been done (it’s much easier to make adjustments in the design phase)
+
+The important thing to understand about the RFCs is that these are NOT a way to request features. Rather, they are a way to bring clarity so people understand what change is being proposed.
```
Maybe add here what we've been saying all along - if you bring an RFC up, it's also assumed that you are proposing to develop it and have necessary resources available at some point in the future :-)

mateusz commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,46 @@
+Once complete submit the RFC as a GitHub issue, in markdown format and include the ID number of the RFC based on the current list of existing proposals.
+
+## What next?
+The RFC will be raised and discussed by the core committers in the monthly Google Hangout sessions, a vote for accepting the RFC will be taken requiring a majority vote (with at least a quorum of more than half of the core committers present).
```
Could also be worth mentioning that the point of this is not finding the best solution, but an acceptable solution. So the submitter still owns the RFC, and core-committers either accept it in full or reject, or propose relatively minor tweaks - this process should not fundamentally change the design of the feature to fit with the opinions of core-committers. If it's acceptable as "a" solution, and it's useful, it should be approved.

camfindlay commented:
Updated this with @mateusz's suggestions. Still think it would be good to have some index of RFC's. I've added other known RFC's (even if they aren't in the prescribed format yet) and a note for people to add as part of the process. If not here then where should we do these?

camfindlay commented:
Note: php has a space to hold all current rfc's https://wiki.php.net/rfc maybe something like this on ss.org?

camfindlay commented:
Be good to get final agreement and merge of this process to help others write good RFCs for the core committer hangout reviews.

jonom commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,52 @@
+
+The rationale behind this process is to:
+ * Encourage visibility on decision making
+ * Clarity on what the proposal is, its rationale and impact
```
Should be: *Clarify what the proposal is, its rationale and impact* or *Provide clarity on what the proposal is, its rationale and impact*

jonom commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,52 @@
+ * Encourage visibility on decision making
+ * Clarity on what the proposal is, its rationale and impact
+ * Reduce unnecessary work when coding large features is done without community and core-committers buy-in
+ * Improved likelihood of an optimal solution being merged
```
Should be: *Improve likelihood of an optimal solution being merged*

jonom commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,52 @@
+ * RFC can be used as a basis for documentation of the feature
+	
+## How to write an RFC?
+### Template
```
This might be a little clearer:
```
The following structure can act as a template for starting your RFC.
 * **Title** - short descriptive title which includes reference number e.g. *RFC-4: Client-side dependency management*.
 * **Metadata** - standardised header containing at least the Author(s), Status, Version and Last Updated fields.
 * **Purpose and outcome** - briefly describe the problem and solution your RFC addresses.
 * **Motivation** - describe the problem in more detail.
 * **Proposal** - describe the solution in more detail, including any questions or considerations that will need to be addressed during development.
 * **Alternatives** - describer any other approaches that were considered during the community discussion phase and why they were not chosen.
 * **Impact** - How will this change potentially impact SilverStripe core? List any positive and negative effects and upgrade considerations.
```

camfindlay replied:
This has been added, thanks @jonom

jonom commented on docs/en/05_Contributing/02_Request_for_comment.md:
```diff
@@ -0,0 +1,52 @@
+### Template
+The following heading can act as a template to starting your RFC.
+ * **Introduction** - include a reference #, title, author
+ * **Metadata** - standardised header containing at least the Author(s), Status and Version fields.
```
Should probably outline what the valid values are for *Status* and their progression e.g. Pending review -> Approved... anything else?

Also the Version field is a bit confusing, especially if you think in SemVer terms. Probably should increase the version number by 1 every time it's edited (whole numbers only, no point releases). Think a Last Updated field should accompany this field since there's no version history or Edited date in GitHub issues and comments.

camfindlay replied:
Have a think about the gist idea. I think it would be good. Just needs some thought about how we would list all the gists. Perhaps using the below table in the docs? The process could be 1. Create the gist 2. Create a PR against docs with your RFC details in the table format. 3. Update the PR as the discussion progresses. 4. Once the final status of the RFC is agreed the PR is merged?

Maybe you could come up with a better way?

The RFC process started by @mateusz was just to trial out the process so we're open to input here still.

This PR is just to start sussing out a workable process we can get going on (and again would be open to change over time).

camfindlay commented:
Great Feedback @jonom I'll keep making updates until we can get to a workable version to adopt. We dont have to get this 100% perfect, just usable. We can always update and improve as we learn more about the process.

jonom commented:
Totally, just wanted to throw a few ideas out - don't want to slow down this getting in to the docs, we're already using the process after all :D

jonom commented:
Oh I didn't see your intro note about PR-ing edits direct to your branch, sorry

camfindlay commented:
@jonom yeap - feel free to send your edits my way. Otherwise I'll have to find some time again to manually make these changes and push to the PR.

jonom commented:
Have opened a PR against your branch @camfindlay

camfindlay commented:
This was discussed in the core committer hangout today https://plus.google.com/b/106943799421369121040/events/csg1b9v8ih9586cah1l1tvd82u4

I will finalise this by creating links to the issue tags for RFCs in favour of the index table at the bottom of the doc.

camfindlay commented:
Will eventually need a merge up to 3.2 and up

camfindlay commented:
@sminnee updated the doc for the RFC process, please feel free to do a final look over and merge at your convenience.

hafriedlander commented:
Some small grammar errors remain, but clear enough. Merging, so further fixes can be raised as further PRs.

hafriedlander commented:
Merged against 3.2 here: https://github.com/silverstripe/silverstripe-framework/commit/9672cd3994011869ddc194a8cc377891220a8d41
